packages/gestalt/src/Dropdown.js

```diff
@@ -0,0 +1,247 @@
+import DropdownContext from './DropdownContextProvider.js';
+import handleContainerScrolling from './utils/keyboardNavigation.js';
+
+type OptionObject = {|
+  label: string,
+  value: string,
+  subtext?: string,
+|};
+type Props = {|
+  anchor?: ?HTMLElement,
+  children: Node,
+  headerContent?: Node,
+  idealDirection?: 'up' | 'right' | 'down' | 'left',
+  onDismiss: () => void,
+  onSelect?: ({|
+    event: SyntheticKeyboardEvent<HTMLElement> | {| keyCode: number |},
+    item: ?OptionObject,
+  |}) => void,
+|};
+
+export default function Dropdown({
+  anchor,
+  children,
+  headerContent,
+  idealDirection = 'down',
+  onDismiss,
+  onSelect,
+}: Props): Node {
+  const getFlattenedChildren = () => {
+    const dropdownChildrenArray = React.Children.toArray(children);
+
+    const items = dropdownChildrenArray.map((child) => {
+      if (
+        child.props.children &&
+        child.type.displayName === 'DropdownSection'
+      ) {
+        return child.props.children;
+      }
+      if (child.type.displayName === 'DropdownItem') {
+        return child;
+      }
+      return null;
+    });
+
+    return items.flat();
+  };
+
+  const availableOptions = getFlattenedChildren();
+  const [hoveredItem, setHoveredItem] = useState<number>(0);
+
+  let selectedElement;
+  const setOptionRef = (optionRef) => {
+    selectedElement = optionRef;
+    if (selectedElement) selectedElement.focus();
+  };
+
+  const containerRef = useRef();
+
+  const handleKeyNavigation = (
+    event: SyntheticKeyboardEvent<HTMLElement> | {| keyCode: number |},
+    direction: -1 | 0 | 1
+  ) => {
+    const newIndex = direction + hoveredItem;
+    const optionsCount = availableOptions.length - 1;
+
+    const KEYS = {
+      ENTER: 0,
+    };
+
+    // If there's an existing item, navigate from that position
+    let cursorIndex = newIndex;
+
+    // If we've reached the end, start at the top
+    if (newIndex > optionsCount) {
+      cursorIndex = 0;
+    }
+    // If we're at the top going backwards, start at the last item
+    else if (newIndex < 0) {
+      cursorIndex = optionsCount;
+    }
+
+    // $FlowFixMe[incompatible-use] flow 0.135.0 upgrade
+    const newItem = availableOptions[cursorIndex].props.option;
+    setHoveredItem(cursorIndex);
+
+    if (direction === KEYS.ENTER) {
+      if (onSelect) onSelect({ event, item: newItem });
+      // $FlowFixMe[incompatible-use] flow 0.135.0 upgrade
+      if (availableOptions[cursorIndex].props.handleSelect)
+        availableOptions[cursorIndex].props.handleSelect({
+          event,
+          item: newItem,
+        });
+    }
+    // Scrolling
+    handleContainerScrolling(direction, containerRef, selectedElement);
+  };
+
+  const handleKeyDown = (event) => {
+    const KEYS = {
+      UP: -1,
+      DOWN: 1,
+      ENTER: 0,
+    };
+    // Up Arrow
+    if (event.keyCode === 38) {
+      handleKeyNavigation(event, KEYS.UP);
+      // $FlowFixMe[prop-missing]
```

Comment:
Yeah, I've been having a lot of issues getting these events typed properly because of Controller's acceptance of just `{| keyCode: number |})` for its event

Reply:
Hm…we should probably discuss more with Christian and Jenny in January. I think we need a better story for our event typing.